### PR TITLE
Add missing docs and enable missing_docs warn lint

### PR DESCRIPTION
**Pull request by antimora**
Added missing pub documents and enable `missing_docs` warn lint for these crates:

- burn
- burn-core
- burn-common
- burn-dataset
- burn-autodiff
- burn-derive

Other crates are still remaining to be fixed, which will be submitted in the next PR

### CODE REVIEW

File: burn-core/src/data/dataloader/base.rs [hunk 1]
```diff
@@ -1,16 +1,24 @@
...
+/// A data loader iterator that can be used to iterate over a data loader.
 pub trait DataLoaderIterator<O>: Iterator<Item = O> {
+    /// Returns the progress of the data loader.
     fn progress(&self) -> Progress;
 }
 
+/// A data loader that can be used to iterate over a dataset.
 pub trait DataLoader<O> {
+    /// Returns the dataset that the data loader is iterating over.
     fn iter<'a>(&'a self) -> Box<dyn DataLoaderIterator<O> + 'a>;
```
Comment (nathanielsimard) [inline burn-core/src/data/dataloader/base.rs:23]: Not quite right

`/// Iterate over the data loader using a boxed [iterator](DataLoaderIterator)`

File: burn-core/src/data/dataloader/builder.rs [hunk 2]
```diff
@@ -26,21 +36,57 @@ where
         }
     }
 
+    /// Sets the batch strategy.
```
Comment (nathanielsimard) [inline burn-core/src/data/dataloader/builder.rs:39]: `Sets the batch size to a fix number. The [fix batch strategy](FixBatchStrategy) will be used.`

File: burn-core/src/nn/cache/base.rs [hunk 3]
```diff
@@ -6,11 +6,17 @@ pub(crate) enum CacheState<T> {
     Empty,
 }
 
+/// A cache for a tensor.
 pub struct TensorCache<B: Backend, const D: usize> {
     pub(crate) state: CacheState<Tensor<B, D>>,
 }
 
 impl<B: Backend, const D: usize> TensorCache<B, D> {
+    /// Gets empty cache.
```
Comment (nathanielsimard) [inline burn-core/src/nn/cache/base.rs:15]: Creates a new empty cache.

**Review by nathanielsimard — APPROVED** [review]
@antimora very minor merge conflict (sorry) to fix and then I'll merge

**Comment by antimora** [comment]
> @antimora very minor merge conflict (sorry) to fix and then I'll merge

OK. No worries. I have fixed the merge conflicts. Ready to be merged.

**Pull request merged**